File: 152. Maximum Product Subarray/maxProduct.java
class Solution {
    public int maxProduct(int[] nums) {
        int max = Integer.MIN_VALUE, iMin = 1, iMax = 1;
        for (int n: nums) {
            if (n < 0) {
                int temp = iMin;
                iMin = iMax;
                iMax = temp;
            }
            iMin = Math.min(iMin * n, n);
            iMax = Math.max(iMax * n, n);
            max = Math.max(max, iMax);
        }
        return max;
    }
}

File: 152. Maximum Product Subarray/maxProduct.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxRes = nums[0]
        iMin = 1
        iMax = 1
        for n in nums:
            if n < 0:
                iMin, iMax = iMax, iMin
            iMin = min(iMin * n, n)
            iMax = max(iMax * n, n)
            maxRes = max(maxRes, iMax)
        return maxRes